FV3/atmos_cubed_sphere/tools/coarse_graining.F90

```diff
@@ -0,0 +1,274 @@
+
+  interface block_sum
+     module procedure block_sum_2d
+  end interface block_sum
+  
+  interface weighted_block_average
+     module procedure weighted_block_average_2d
+     module procedure weighted_block_average_3d_field_2d_weights
+  end interface weighted_block_average
+  
+  ! Global variables for the module, initialized in coarse_graining_init
+  type(domain2d) :: coarse_domain
+  integer :: is, ie, js, je, npz
+  integer :: is_coarse, ie_coarse, js_coarse, je_coarse
+  integer :: nx_coarse
+  character(len=11) :: MODEL_LEVEL = 'model_level'
+  
+  ! Namelist parameters initialized with default values
+  integer :: coarsening_factor = 8
+  integer :: coarse_io_layout(2) = (/1, 1/)
+  character(len=64) :: strategy = 'model_level'  ! Valid values are 'model_level'
+  logical :: do_coarse_graining = .false.
+  
+  namelist /coarse_graining_nml/ coarsening_factor, coarse_io_layout, strategy, do_coarse_graining
+
+contains
+
+  subroutine coarse_graining_init(npx, atm_npz, layout, bd, coarse_graining)
+    integer, intent(in) :: npx
+    integer, intent(in) :: atm_npz
+    integer, intent(in) :: layout(2)
+    type(fv_grid_bounds_type), intent(in) :: bd
+    type(fv_coarse_graining_type), intent(inout) :: coarse_graining
+
+    character(len=256) :: error_message
+    logical :: exists
+    integer :: error_code, iostat
+
+    read(input_nml_file, coarse_graining_nml, iostat=iostat)
+    error_code = check_nml_error(iostat, 'coarse_graining_nml')
+    coarse_graining%do_coarse_graining = do_coarse_graining
+
+    if (do_coarse_graining) then
+       call assert_valid_strategy(strategy)
+       call compute_nx_coarse(npx, coarsening_factor, nx_coarse)
+       call assert_valid_domain_layout(nx_coarse, layout)
+       call define_cubic_mosaic(coarse_domain, nx_coarse, nx_coarse, layout)
+       call mpp_define_io_domain(coarse_domain, coarse_io_layout)
+       call mpp_get_compute_domain(coarse_domain, is_coarse, ie_coarse, js_coarse, je_coarse)
+       call get_fine_array_bounds(bd, is, ie, js, je)
+       npz = atm_npz
+       
+       coarse_graining%bd%is_coarse = is_coarse
+       coarse_graining%bd%ie_coarse = ie_coarse
+       coarse_graining%bd%js_coarse = js_coarse
+       coarse_graining%bd%je_coarse = je_coarse
+       coarse_graining%nx_coarse = nx_coarse
+       coarse_graining%domain = coarse_domain
+       coarse_graining%strategy = strategy
+    endif    
+  end subroutine coarse_graining_init
+
+  subroutine compute_nx_coarse(npx, coarsening_factor, nx_coarse)
+    integer, intent(in) :: npx
+    integer, intent(in) :: coarsening_factor
+    integer, intent(out) :: nx_coarse
+
+    character(len=256) :: error_message
+    integer :: nx
+    
+    nx = npx - 1
+    if (mod(nx, coarsening_factor) > 0) then
+       write(error_message, *) 'coarse_graining_init: coarsening_factor does not evenly divide the native resolution.'
+       call mpp_error(FATAL, error_message)
+    endif
+    nx_coarse = nx / coarsening_factor
```

Comment:
Make sure you don't get burned in other places by `npx` being interface levels while `nx_coarse` is cell centers.